### ISSUE TITLE
Match statement incorrectly flagged as inexhaustive due to suggested abstract base class

### ISSUE THREAD

**Issue by danr** [microsoft/pyright#2945]
**Describe the bug**
Abstract base class suggested as case pattern in match.

**To Reproduce**
```
$ cat test.py
from dataclasses import dataclass
import abc

class Command(abc.ABC):
    pass

@dataclass(frozen=True)
class PrintGreeting(Command):
    greeting: str

def execute(command: Command):
    match command:
        case PrintGreeting():
            print(command.greeting)
$ pyright test.py
Assuming Python platform Linux
Auto-excluding **/node_modules
Auto-excluding **/__pycache__
Auto-excluding .git
stubPath /tmp/z/typings is not a valid directory.
Searching for source files
File or directory "/tmp/z not exist.
Found 1 source file
/tmp/z/test.py
  /tmp/z/test.py:12:11 - error: Cases within match statement do not exhaustively handle all values
    Unhandled type: "Command"
    If exhaustive handling is not intended, add "case _: pass" (reportMatchNotExhaustive)
1 error, 0 warnings, 0 informations
Completed in 0.571sec
```

**Expected behaviour**
Pyright should not suggest `Command` as a case candidate as it is abstract.

**Additional context**
```
$ pyright --version
pyright 1.1.215
```

**Comment by danr** [comment]
Apparently having ABC as a base class does not prevent it a class from being instatiated.
This runs without an issue:

```python
class Command2(abc.ABC):
    pass

c2 = Command2()
```

However, adding an abstract method gives a type error:

```python
class Command3(abc.ABC):
    @abc.abstractmethod
    def m(self):
        pass

c3 = Command3()
```

```
Traceback (most recent call last):
  File "/tmp/z/test.py", line 24, in <module>
    c3 = Command3()
TypeError: Can't instantiate abstract class Command3 with abstract method m
```

I'm just looking for a way pyright accepts to encode algebraic data types on top of dataclasses.

**Comment by danr** [comment]
Thanks for your reply. I realized I thought the class would somehow be sealed for extensions in "later" code, such as ther modules, but classes in python are of course always open for subclassing. It would be useful to have a way to have closed classes for algebraic data types.

**Comment by erictraut** [comment]
You can close a class by marking it `@final`, but it wouldn't make sense to mark an ABC class as `@final` because that wouldn't allow any other class to subclass from it.

**Comment by erictraut** [comment]
Another technique that is sometimes useful is to define a type alias that refers to a union of specified types. 

```python
Commands = PrintGreeting | PlayDead | DoSomethingElse
```
